### PR TITLE
[clang] Implement TTP P0522 pack matching for deduced function template calls.

### LINKED ISSUE #111363: Clang++ fails to match template template parameters as function arguments
Clang++ trunk started rejecting the following testcase (adapted from https://eel.is/c++draft/temp#arg.template-example-3). Clang 19 accepts it. It's a recent regression.

```c++
template <template <class, class...> class TT, class T1, class... Rest>
void f(TT<T1, Rest...> const &) {}

template <typename T1> struct A {};

int main()
{ 
  A<int> a;
  f(a);   // bogus "no matching function for call to 'f'"
}
```

### PR DESCRIPTION
**Pull request by mizvekov**
Clang previously missed implementing P0522 pack matching for deduced function template calls.

Fixes #111363

### CODE REVIEW

File: clang/include/clang/Sema/Overload.h [hunk 1]
```diff
@@ -925,6 +925,8 @@ class Sema;
 
     bool TookAddressOfOverload : 1;
 
+    bool HasMatchedPackOnParmToNonPackOnArg : 1;
```
Comment (cor3ntin) [inline clang/include/clang/Sema/Overload.h:928]: Can you add a comment?

File: clang/lib/Sema/SemaTemplateDeduction.cpp [hunk 2]
```diff
@@ -3172,10 +3172,11 @@ static TemplateDeductionResult ConvertDeducedTemplateArguments(
     }
 
     // Check whether we can actually use the default argument.
-    if (S.CheckTemplateArgument(
-            Param, DefArg, TD, TD->getLocation(), TD->getSourceRange().getEnd(),
-            0, SugaredBuilder, CanonicalBuilder, Sema::CTAK_Specified,
-            /*MatchedPackOnParmToNonPackOnArg=*/nullptr)) {
+    if (S.CheckTemplateArgument(Param, DefArg, TD, TD->getLocation(),
+                                TD->getSourceRange().getEnd(), 0,
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaTemplateDeduction.cpp:3176]: can we add a comment for 0?

**Review by cor3ntin — APPROVED** [review]
LGTM.
I don't think we need a changelog, the one from P0522 ought to be enough (we can add details to it once everything is merged)

**Comment by mizvekov** [comment]
@zmodem FIY this will fix https://github.com/llvm/llvm-project/pull/96023#issuecomment-2393228464

**Comment by zmodem** [comment]
Great!

**Pull request merged**